lightning/src/ln/peers/handler.rs

```diff
@@ -0,0 +1,3280 @@
+use ln::channelmanager::{SimpleArcChannelManager, SimpleRefChannelManager};
+use util::ser::{Writeable};
+use ln::wire;
+use ln::wire::{Encode, Message};
+use util::byte_utils;
+use util::events::{MessageSendEvent, MessageSendEventsProvider};
+use util::logger::Logger;
+use routing::network_graph::NetGraphMsgHandler;
+
+use std::collections::HashMap;
+use std::sync::{Arc, Mutex};
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::{cmp,error,hash,fmt};
+use std::ops::Deref;
+
+use bitcoin::hashes::sha256::Hash as Sha256;
+use bitcoin::hashes::sha256::HashEngine as Sha256Engine;
+use bitcoin::hashes::{HashEngine, Hash};
+use ln::peers::outbound_queue::OutboundQueue;
+use ln::peers::transport::{PayloadQueuer, Transport};
+use bitcoin::hashes::core::iter::Filter;
+use std::collections::hash_map::IterMut;
+
+// Number of items that can exist in the OutboundQueue before Sync message flow control is triggered
+const OUTBOUND_QUEUE_SIZE: usize = 10;
+
+/// Interface PeerManager uses to interact with the Transport object
+pub(super) trait ITransport: MessageQueuer {
+	/// Instantiate the new outbound Transport
+	fn new_outbound(initiator_static_private_key: &SecretKey, responder_static_public_key: &PublicKey, initiator_ephemeral_private_key: &SecretKey) -> Self;
+
+	/// Set up the Transport receiving any bytes that need to be sent to the peer
+	fn set_up_outbound(&mut self) -> Vec<u8>;
+
+	/// Instantiate a new inbound Transport
+	fn new_inbound(responder_static_private_key: &SecretKey, responder_ephemeral_private_key: &SecretKey) -> Self;
+
+	/// Process input data similar to reading it off a descriptor directly. Returns true on the first call
+	/// that results in the transport being newly connected.
+	fn process_input(&mut self, input: &[u8], output_buffer: &mut impl PayloadQueuer) -> Result<bool, String>;
+
+	/// Returns true if the connection is established and encrypted messages can be sent.
+	fn is_connected(&self) -> bool;
+
+	/// Returns the node_id of the remote node. Panics if not connected.
+	fn get_their_node_id(&self) -> PublicKey;
+
+	/// Returns all Messages that have been received and can be successfully parsed by the Transport
+	fn drain_messages<L: Deref>(&mut self, logger: L) -> Result<Vec<Message>, PeerHandleError> where L::Target: Logger;
+}
+
+/// Interface PeerManager uses to queue message to send. Implemented by Transport to handle
+/// encryption/decryption post-NOISE.
+pub(super) trait MessageQueuer {
+	/// Encodes, encrypts, and enqueues a message to the outbound queue. Panics if the connection is
+	/// not established yet.
+	fn enqueue_message<M: Encode + Writeable, Q: PayloadQueuer, L: Deref>(&mut self, message: &M, output_buffer: &mut Q, logger: L) where L::Target: Logger;
+}
+
+/// Trait representing a container that can try to flush data through a SocketDescriptor. Used by the
+/// PeerManager to handle flushing the outbound queue and flow control.
+pub(super) trait SocketDescriptorFlusher {
```

Comment:
Separating what an object can do and what another object needs it to provide allows for a more modular design and easier testing. The fact that `OutboundQueue` implements both interfaces is more of a convenience than a reason to flattern the interfaces.

By splitting the entire API into the interfaces that each object needs, you can easily write tests by just implementing test doubles for the subset of the API and not have to mock/spy a bunch of unused features/function. It also allows the function signatures to document the operations that the function will use which can come in handy when trying to restrict operations in different phase (such as no flush in `read_event()`)

### Testing
The transport tests for `process_input()` only need access to the subset of the API that enqueues messages for send. This allows the tests to use a simple Vec<u8> spy instead of a full OutboundQueue and get the same test coverage.

### Interface vs Object Development
Using interfaces instead of objects also focuses the attention on which features a consumer object needs to get their work done. In my experience, this leads to smaller interfaces between objects that don't overengineer anything.

Another approach I experimented with was to have a separate `SocketDescriptorFlusherImpl` object that uses composition w/ a `PayloadQueuer` and just handles the logic to add Sync messages and flow control. This separate object would implement the `SocketDescriptorFlusher` trait and fits more inline with the single responsibility principle.

Given the longer feedback loops for open source work and my experience with OO principles being less engrained in open source projects, I chose to limit the number of new objects and first see if this was an approach that was interesting to the project.